File: discover_javascript/github_locations.js
const https = require('https');
const fs = require('fs');

const options = {
	hostname: 'api.github.com',
	path: '/search/repositories?q=language:javascript&sort=stars&order=desc',
	headers: {
	'User-Agent': 'Holberton_School',
	'Authorization': 'token 6a54def2525aa32b003337b31487e321d6a2bb59'
  }
}

var streamToString = (stream, cb) => {
  const chunks = [];
  stream.on('data', (chunk) => {
    chunks.push(chunk);
  });
  stream.on('end', () => {
    cb(chunks.join(''));
  });
}

const req = https.request(options, (res) => {
    console.time("dbsave");
	streamToString(res, (data) => {
		const gitObj = JSON.parse(data);
        const result = [];
        const items = gitObj.items;
        items.map((obj) =>{
            result.push({fullName: obj.full_name ,login: obj.owner.login});
        })
        compute(result);
	});
});
req.end();

req.on('error', (e) => {
	console.log(e);
});

const compute = (result) => {
    const final = [];
    result.map((user) => {
        options.path = '/users/'+user.login;
        const reqUser = https.request(options, (res) => {
            streamToString(res, (data) => {
                data = JSON.parse(data);
                final.push({"full_name": user.fullName, "location": data.location});
		console.log(user.fullName);
                complete();
            });
        });
        reqUser.end();
        reqUser.on('error', (e) => {
            console.log(e);
        });
    });
    const complete = () => {
        if(final.length == 30){
	    console.log(JSON.stringify(final));
	    console.timeEnd("dbsave");
            /*fs.writeFile("/tmp/23", JSON.stringify(final), (err) => {
                if(err) {
                    console.log(err);
                }
                console.log("The file was saved!");
            });*/
        } 
    }
};

File: discover_perl/github_locations.pl
#!/usr/bin/perl

=pod
* ===---Description------------------------------------------------------------===
*   Github Perl most popular Projects, based on Rudy Rigoy projects.
*   Using a Perl script to get the 30 most popular Perl Github repositories.
*
*   by Alexandro de Oliveira, for Holberton School
* ===--------------------------------------------------------------------------===
=cut

use strict;
use warnings;
use LWP::UserAgent;
use Data::Dump "pp";            # Used to print and analize complex objects.
use JSON;

my $url = 'https://api.github.com/search/repositories?q=language:perl&sort=stars&order=desc';

my $headers = HTTP::Headers->new(
'User-Agent'=> 'Holberton_School',
'Authorization' => 'token 6a54def2525aa32b003337b31487e321d6a2bb59',
);

my $ua = LWP::UserAgent->new(ssl_opts => { verify_hostname => 0}); # bypass SSL authentication checks.

my $req = HTTP::Request->new(GET => $url, $headers);
# Print the request object:
# print pp($req);

# Requesting/Connecting:
my $res = $ua->request($req);
# print pp($res);
die "Error getting $url\n", $res->status_line
unless $res->is_success();

my $myjson = decode_json($res->content());
my @result;
foreach my $item (@{$myjson -> {items}})
{
    my $login = $item->{owner}->{login};
    my $full_name = $item->{full_name};
    my $location = fetchData($login);
    my %obj = (full_name => $login, location => $location);
    push @result, \%obj;
}

print encode_json(\@result);

# Fetching the location of each user:
sub fetchData
{
    my $login = shift; # Defining the parameter to this subroutine.
    my $url = "https://api.github.com/users/" . $login;
    my $ua = LWP::UserAgent->new(ssl_opts => {verify_hostname => 0 });
    my $req = HTTP::Request->new( GET => $url, $headers);
    my $res = $ua->request($req);
    die "Error requesting $login $!\n", $res->status_line
    unless $res->is_success();
    my $data = decode_json($res->content);
    return $data->{location};
}
